src/app/pages/admin-page/admin-page.component.ts

```diff
@@ -1,8 +1,290 @@
+      if (emptyControls.length > 1) {
+        this.deleteRelation(changedIndex);
+      }
+    }
+  }
+
+  protected onSubmit(): void {
+    if (this.stationForm.valid) {
+      const formValue = this.stationForm.value as {
+        city: string;
+        latitude: number;
+        longitude: number;
+        relations: number[];
+      };
+      this.store.dispatch(
+        createStation({
+          city: formValue.city,
+          latitude: formValue.latitude,
+          longitude: formValue.longitude,
+          relations: formValue.relations,
+        }),
+      );
+      if (this.marker) {
+        this.map.removeLayer(this.marker);
+      }
+      this.resetForm();
+    }
+  }
+
+  private resetForm() {
+    this.stationForm.reset({
+      city: '',
+      latitude: 0,
+      longitude: 0,
+    });
+    this.resetRelations();
+  }
+
+  protected updateOptions() {
+    const selectedValues = this.relations.controls.map((control) => {
+      return control.value;
+    });
+    this.relations.controls.forEach((control) => {
+      control.setValidators([Validators.required, this.forbidSelectedValues(selectedValues)]);
+      control.updateValueAndValidity();
+    });
+  }
+
+  private resetRelations() {
+    const relationsArray = this.stationForm.get('relations') as FormArray;
+    while (relationsArray.length) {
+      relationsArray.removeAt(0);
+      const subscription = this.formGroupSubscriptions[0];
+      if (subscription) {
+        subscription.unsubscribe();
+      }
+    }
+    relationsArray.push(this.createRelation(0));
+  }
+
+  private forbidSelectedValues(selectedValues: string[]) {
+    return (control: AbstractControl) => {
+      return selectedValues.includes(control.value)
+        ? { forbiddenValue: { value: control.value } }
+        : null;
+    };
+  }
+
+  protected isOptionSelected(option: number): boolean {
+    const selectedValues = this.relations.controls.map((control) => {
+      return control.value;
+    });
+    return selectedValues.includes(option);
+  }
+
+  protected onDelete(id: number, stations: Station[]) {
+    const station = stations.find((s) => {
+      return s.id === id;
+    });
+    if (!station) {
+      throw new Error('Station not found');
+    }
+
+    const coordinates = station.connectedTo.map((connection) => {
+      const connectedStation = stations.find((s) => {
+        return s.id === connection.id;
+      });
+      if (!connectedStation) {
+        throw new Error(`Connected station with id ${connection.id} not found`);
+      }
+      return {
+        latitude: connectedStation.latitude,
+        longitude: connectedStation.longitude,
+      };
+    });
+    this.store.dispatch(canDelete({ station, coordinates }));
+  }
+
+  protected findStationNameById(id: number, stations: Station[]): string {
+    const foundStation = stations.find((station) => {
+      return station.id === id;
+    });
+    return foundStation ? foundStation.city : 'Station not found';
+  }
+}
```

Comment:
Please move the Stations to a separate component to insert into the Admin Panel page.

Reply:
Okay, i fixed that